### PR TITLE
Loads .pyd instead of .so in MemPool test for windows

### LINKED ISSUE #132650: mempool test fails on windows
The new feature added in https://github.com/pytorch/pytorch/pull/131152 fails on windows cuda CI (that runs periodic only).
You can see an example failure here: https://hud.pytorch.org/pytorch/pytorch/commit/27f61eba58572ffc0f3ac766d07d69ca31947ab6

```
2024-08-05T10:49:48.9034514Z ___________________ TestMemPool.test_mempool_with_allocator ___________________
2024-08-05T10:49:48.9034928Z Traceback (most recent call last):
2024-08-05T10:49:48.9035618Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_cuda.py", line 5009, in test_mempool_with_allocator
2024-08-05T10:49:48.9036286Z     allocator = torch.cuda.memory.CUDAPluggableAllocator(
2024-08-05T10:49:48.9036862Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\cuda\memory.py", line 929, in __init__
2024-08-05T10:49:48.9037383Z     allocator = ctypes.CDLL(path_to_so_file)
2024-08-05T10:49:48.9037769Z   File "C:\Jenkins\Miniconda3\lib\ctypes\__init__.py", line 374, in __init__
2024-08-05T10:49:48.9038163Z     self._handle = _dlopen(self._name, mode)
2024-08-05T10:49:48.9038957Z FileNotFoundError: Could not find module 'C:\Users\RUNNER~1\AppData\Local\Temp\tmp5ev_rqkd\dummy_allocator.so' (or one of its dependencies). Try using the full path with constructor syntax.
2024-08-05T10:49:48.9039645Z 
2024-08-05T10:49:48.9039886Z During handling of the above exception, another exception occurred:
2024-08-05T10:49:48.9040177Z 
2024-08-05T10:49:48.9040268Z Traceback (most recent call last):
2024-08-05T10:49:48.9040627Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 627, in _rmtree_unsafe
2024-08-05T10:49:48.9040978Z     os.unlink(fullname)
2024-08-05T10:49:48.9041491Z PermissionError: [WinError 5] Access is denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp5ev_rqkd\\dummy_allocator_v2.pyd'
2024-08-05T10:49:48.9041969Z 
2024-08-05T10:49:48.9042150Z During handling of the above exception, another exception occurred:
2024-08-05T10:49:48.9042433Z 
2024-08-05T10:49:48.9042529Z Traceback (most recent call last):
2024-08-05T10:49:48.9042867Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 814, in onerror
2024-08-05T10:49:48.9043259Z     _os.unlink(path)
2024-08-05T10:49:48.9043758Z PermissionError: [WinError 5] Access is denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp5ev_rqkd\\dummy_allocator_v2.pyd'
2024-08-05T10:49:48.9044227Z 
2024-08-05T10:49:48.9044409Z During handling of the above exception, another exception occurred:
2024-08-05T10:49:48.9044696Z 
2024-08-05T10:49:48.9044783Z Traceback (most recent call last):
2024-08-05T10:49:48.9045268Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_cuda.py", line 5017, in test_mempool_with_allocator
2024-08-05T10:49:48.9045828Z     self.assertEqual(allocator.allocator(), pool.allocator)
2024-08-05T10:49:48.9046262Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 839, in __exit__
2024-08-05T10:49:48.9046611Z     self.cleanup()
2024-08-05T10:49:48.9046901Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 843, in cleanup
2024-08-05T10:49:48.9047252Z     self._rmtree(self.name)
2024-08-05T10:49:48.9047571Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 825, in _rmtree
2024-08-05T10:49:48.9047934Z     _shutil.rmtree(name, onerror=onerror)
2024-08-05T10:49:48.9048286Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 759, in rmtree
2024-08-05T10:49:48.9048640Z     return _rmtree_unsafe(path, onerror)
2024-08-05T10:49:48.9049067Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 629, in _rmtree_unsafe
2024-08-05T10:49:48.9049459Z     onerror(os.unlink, fullname, sys.exc_info())
2024-08-05T10:49:48.9049838Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 817, in onerror
2024-08-05T10:49:48.9050178Z     cls._rmtree(path)
2024-08-05T10:49:48.9050472Z   File "C:\Jenkins\Miniconda3\lib\tempfile.py", line 825, in _rmtree
2024-08-05T10:49:48.9050835Z     _shutil.rmtree(name, onerror=onerror)
2024-08-05T10:49:48.9051180Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 759, in rmtree
2024-08-05T10:49:48.9051523Z     return _rmtree_unsafe(path, onerror)
2024-08-05T10:49:48.9051890Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 610, in _rmtree_unsafe
2024-08-05T10:49:48.9052277Z     onerror(os.scandir, path, sys.exc_info())
2024-08-05T10:49:48.9052653Z   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 607, in _rmtree_unsafe
2024-08-05T10:49:48.9053028Z     with os.scandir(path) as scandir_it:
2024-08-05T10:49:48.9053699Z NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp5ev_rqkd\\dummy_allocator_v2.pyd'
2024-08-05T10:49:48.9054660Z - generated xml file: C:\actions-runner\_work\pytorch\pytorch\test\test-reports\python-pytest\test_cuda\test_cuda-8f7e3caa91a782f5.xml -
2024-08-05T10:49:48.9055315Z =========================== short test summary info ===========================
2024-08-05T10:49:48.9055771Z FAILED [13.1202s] test_cuda.py::TestMemPool::test_mempool_with_allocator - No...
2024-08-05T10:49:48.9056362Z !!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
2024-08-05T10:49:48.9057038Z ================= 1 failed, 182 deselected, 2 rerun in 42.73s =================
```

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @ptrblck @msaroufim

### PR DESCRIPTION
**Pull request by syed-ahmed**
Fixes #132650

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -4994,9 +4993,11 @@ def test_mempool_with_allocator(self):
         from torch.utils.cpp_extension import load_inline
 
         dummy_allocator_source = """
+        #include <torch/extension.h>
         extern "C" {
-          void* dummy_alloc(size_t size, int device, void* stream) { return nullptr; }
-          void dummy_free(void* ptr) { }
+          // Note that windows needs __declspec(dllexport): https://stackoverflow.com/a/24575865
+          TORCH_API void* dummy_alloc(size_t size, int device, void* stream) { return nullptr; }
```
Comment (albanD) [inline test/test_cuda.py:4999]: I am not sure TORCH_API will work here as it depends on the build flags. Maybe C10_EXPORT would be better?

**Review by albanD — APPROVED** [review]
Thanks for the quick fix!

File: test/test_cuda.py (hunk 1 above)
Comment (syed-ahmed) [reply test/test_cuda.py]: Thank you! Like like C10_EXPORT was the right thing to use.

**Comment by syed-ahmed** [comment]
I checked all the shards of `periodic / win-vs2019-cuda11.8-py3 / test` and none of them had any mempool failures. Seems like two things that solved the issue was:
- I was loading .so instead of .pyd in windows which was causing a library not found error. So retrieving the correct path for the .pyd and loading that in CUDAPluggableAllocator solved it.
- Also the allocator functions weren't exposed in the dll, so ctypes couldn't find them. Adding C10_EXPORT solved that issue. 

So gonna ahead and merge it.

**Comment by syed-ahmed** [comment]
`PermissionError: [WinError 5] Access is denied:` was fixed in the latest commit by not using `tempfile`. Tempfile in windows has permission issues with closing and deleting files. So the latest commit uses the existing clean up logic in `load_inline` (`keep_intermediates=False`) for the failing test. The passing test can be seen in the artifacts of `win-vs2019-cuda11.8-py3 / test (default, 2, 4, windows.g5.4xlarge.nvidia.gpu)` at https://hud.pytorch.org/pr/132749. So merging again on this basis.

**Pull request closed without merging**